vortex-array/src/parts.rs

```diff
@@ -74,3 +79,102 @@ impl ArrayParts {
         )
     }
 }
+
+/// Convert an [`ArrayData`] into [`ArrayParts`].
+impl From<ArrayData> for ArrayParts {
+    fn from(array: ArrayData) -> Self {
+        let flatbuffer = ArrayPartsFlatBuffer {
+            array: &array,
+            buffer_idx: 0,
+        }
+        .write_flatbuffer_bytes();
+        let mut buffers: Vec<ByteBuffer> = vec![];
+        for child in array.depth_first_traversal() {
+            for buffer in child.byte_buffers() {
+                buffers.push(buffer);
+            }
+        }
+        Self {
+            row_count: array.len(),
+            flatbuffer,
+            flatbuffer_loc: 0,
+            buffers,
+        }
+    }
+}
+
+/// A utility struct for creating an [`fba::Array`] flatbuffer.
+pub struct ArrayPartsFlatBuffer<'a> {
+    array: &'a ArrayData,
+    buffer_idx: u16,
+}
+
+impl<'a> ArrayPartsFlatBuffer<'a> {
+    pub fn new(array: &'a ArrayData) -> Self {
+        Self {
+            array,
+            buffer_idx: 0,
+        }
+    }
+}
+
+impl FlatBufferRoot for ArrayPartsFlatBuffer<'_> {}
+
+impl WriteFlatBuffer for ArrayPartsFlatBuffer<'_> {
+    type Target<'t> = fba::Array<'t>;
+
+    fn write_flatbuffer<'fb>(
+        &self,
+        fbb: &mut FlatBufferBuilder<'fb>,
+    ) -> WIPOffset<Self::Target<'fb>> {
+        let encoding = self.array.encoding().id().code();
+        let metadata = self
+            .array
+            .metadata_bytes()
+            .vortex_expect("IPCArray is missing metadata during serialization");
+        let metadata = Some(fbb.create_vector(metadata.as_ref()));
+
+        // Assign buffer indices for all child arrays.
+        // The second tuple element holds the buffer_index for this Array subtree. If this array
+        // has a buffer, that is its buffer index. If it does not, that buffer index belongs
+        // to one of the children.
+        let nbuffers = u16::try_from(self.array.nbuffers())
+            .vortex_expect("Array can have at most u16::MAX buffers");
+        let child_buffer_idx = self.buffer_idx + nbuffers;
+
+        let children = self
+            .array
+            .children()
+            .iter()
+            .scan(child_buffer_idx, |buffer_idx, child| {
+                // Update the number of buffers required.
+                let msg = ArrayPartsFlatBuffer {
+                    array: child,
+                    buffer_idx: *buffer_idx,
+                }
+                .write_flatbuffer(fbb);
+                *buffer_idx = u16::try_from(child.cumulative_nbuffers())
+                    .ok()
+                    .and_then(|nbuffers| nbuffers.checked_add(*buffer_idx))
+                    .vortex_expect("Too many buffers (u16) for ArrayData");
+                Some(msg)
+            })
+            .collect_vec();
```

Comment:
fwiw I found that scan behaves poorly since it doesn't know the size of the iterator so it will grow the collected vec one element at the time. Since in this case we always return Some and never None I think it's better to preallocate the vec and have a loop

Reply:
This was pre-existing code just moved around. I agree, just might be a FLUP